### PR TITLE
Bugfixes for Event Structure model checking strategy

### PR DESCRIPTION
**Pull request by viliamDokov**
This enables event structure for most tests in `lincheck_test.datastructures`. 
It also includes a lot of small bugfixes.
The most notable feature is the support of `System.arraycopy`.  

Note for @eupp, this does not have the same results, as in the Evaluation as some tests time out.
Also the fixes from the constructor instrumentation branch are also not included.  

I will clean up the optmization and PR and send it over as well.

### CODE REVIEW

**Comment by viliamDokov** [comment]
I actually see that the branch has been merged. I will rebase it and add tests for that as well.

File: integration-test/lincheck/build.gradle.kts [hunk 1]
```diff
@@ -16,6 +16,14 @@ sourceSets {
     }
 }
 
+plugins {
```
Comment (viliamDokov) [inline integration-test/lincheck/build.gradle.kts:19]: NOTE: This change is part of this slack discussion: 
https://soham-jetbrains.slack.com/archives/D09TVFBD286/p1781094892568339

File: lincheck/src/jvm/main/org/jetbrains/kotlinx/lincheck/execution/RandomExecutionGenerator.java [hunk 2]
```diff
@@ -96,7 +97,9 @@ public ExecutionScenario nextExecution() {
         } else {
             postExecution = Collections.emptyList();
         }
-        return new ExecutionScenario(initExecution, parallelExecution, postExecution, testStructure.validationFunction);
+        ExecutionScenario result = new ExecutionScenario(initExecution, parallelExecution, postExecution, testStructure.validationFunction);
+        System.out.println("Generated execution" + (++i) + ":\n" + result);
+        return result;
```
Comment (viliamDokov) [inline lincheck/src/jvm/main/org/jetbrains/kotlinx/lincheck/execution/RandomExecutionGenerator.java]: remove debugging

File: lincheck/src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/eventstructure/EventStructureStrategy.kt [hunk 3]
```diff
@@ -690,8 +702,8 @@ private class EventStructureMemoryTracker(
     }
 
     override fun interceptArrayCopy(iThread: Int, codeLocation: Int, srcArray: Any?, srcPos: Int, dstArray: Any?, dstPos: Int, length: Int) {
-        val srcType = srcArray!!::class.getType()
-        val dstType = dstArray!!::class.getType()
+        val srcType = srcArray!!.javaClass.kotlin.getArrayElementType()
```
Comment (viliamDokov) [inline lincheck/src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/eventstructure/EventStructureStrategy.kt]: This change was an annoying bug. Maybe there should be a nice function where you pass in an array object and return the element type, without any confusion?

File: lincheck/src/jvm/test-jdk11/org/jetbrains/kotlinx/lincheck_test/strategy/eventstructure/VarHandleTests.kt [hunk 4]
```diff
@@ -201,4 +212,50 @@ class VarHandleTests {
         }
     }
 
+
+    class VarHandleTestClassWithInnerNode {
+
```
Comment (viliamDokov) [inline lincheck/src/jvm/test-jdk11/org/jetbrains/kotlinx/lincheck_test/strategy/eventstructure/VarHandleTests.kt:217]: Verify if this test actually works

File: lincheck/src/jvm/test/org/jetbrains/kotlinx/lincheck_test/strategy/eventstructure/PrimitivesTest.kt [hunk 5]
```diff
@@ -1467,6 +1476,333 @@ class PrimitivesTest {
...
+            t1.join()
+            t2.join()
+            t3.join()
+            r0
+        }
+    }
+
+    @Test
+    fun testArrayCopy() {
+        val outcomes = setOf(
```
Comment (viliamDokov) [inline lincheck/src/jvm/test/org/jetbrains/kotlinx/lincheck_test/strategy/eventstructure/PrimitivesTest.kt]: A lot of repetition for these tests. Either we delete the string/box ones or I make some sort of Util function so the parametrization is nicer.

File: lincheck/src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/eventstructure/EventStructureStrategy.kt (hunk 3 above)
Comment (eupp) [reply lincheck/src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/eventstructure/EventStructureStrategy.kt]: Yes, feel free to add such function somewhere in `common/utils`.

File: bootstrap/src/sun/nio/ch/lincheck/EventTracker.java [hunk 6]
```diff
@@ -54,6 +54,8 @@ public interface EventTracker {
     void afterLocalRead(ThreadDescriptor descriptor, int codeLocation, int variableId, Object value);
     void afterLocalWrite(ThreadDescriptor descriptor, int codeLocation, int variableId, Object value);
 
+    void onArrayCopy(ThreadDescriptor descriptor, Object srcArray, int srcPos, Object dstArray, int dstPos, int length);
```
Comment (eupp) [inline bootstrap/src/sun/nio/ch/lincheck/EventTracker.java]: As I understand, you probably took part of this code regarding `onArrayCopy` instrumentation from an old branch, right?

Now Lincheck uses a bit different way to detect `Arrays.copyOf(...)` calls that does not need special instrumentation and injection method; it is just detected via regular `onMethodCall` injection. 
You can have a look at how it is done in `ManagedStrategy.processIntrinsicMethodEffects` method. 

Could you please try to use the same apprach?
That would greatly simplify the code, as we would not need new injection method and the new bytecode transformer `ArrayTransformer`.

File: integration-test/lincheck/src/main/org/jetbrains/lincheck_test/datastructures/linearizable/BufferedChannelTest.kt [hunk 7]
```diff
@@ -33,6 +36,8 @@ class BufferedChannelTest : AbstractLincheckTest() {
     fun poll() = c.tryReceive().getOrNull()
 
     override fun <O : Options<O, *>> O.customize() {
+        // TODO: We ingore for now since event structure strategy does not support coroutines.
+        assumeTrue(this !is ModelCheckingOptions || !isExperimentalModelCheckingEnabled)
```
Comment (eupp) [inline integration-test/lincheck/src/main/org/jetbrains/lincheck_test/datastructures/linearizable/BufferedChannelTest.kt]: I think you have `ignoreExperimentalModelChecking` extension function just for this purpose?

File: integration-test/lincheck/src/main/org/jetbrains/lincheck_test/datastructures/linearizable/ConcurrentHashMapTest.kt [hunk 8]
```diff
@@ -32,6 +32,7 @@ class ConcurrentHashMapTest : AbstractLincheckTest() {
     fun remove(@Param(name = "key") key: Int) = map.remove(key)
 
     override fun <O : Options<O, *>> O.customize() {
+        ignoreExperimentalModelChecking()
```
Comment (eupp) [inline integration-test/lincheck/src/main/org/jetbrains/lincheck_test/datastructures/linearizable/ConcurrentHashMapTest.kt:36]: Please add a TODO comment to explain why it is ignored.

File: integration-test/lincheck/src/main/org/jetbrains/lincheck_test/guide/MPSCQueueTest.kt [hunk 9]
```diff
@@ -33,4 +34,8 @@ class MPSCQueueTest {
 
     @Test
     fun modelCheckingTest() = ModelCheckingOptions().check(this::class)
+
+    @Ignore("We need to add loop detectors for this test")
```
Comment (eupp) [inline integration-test/lincheck/src/main/org/jetbrains/lincheck_test/guide/MPSCQueueTest.kt]:
```suggestion
    @Ignore("We need to support loop detection for this test")
```

File: lincheck/src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/eventstructure/EventStructureObjectTracker.kt [hunk 10]
```diff
@@ -21,14 +21,18 @@
...
 import org.jetbrains.lincheck.util.toInt
 import sun.nio.ch.lincheck.TestThread
 import java.lang.ref.WeakReference
 
 
-internal class EventStructureObjectTracker(private val eventStructure: EventStructure): BaseObjectTracker() {
+internal class EventStructureObjectTracker(
+    private val eventStructure: EventStructure,
+    private val currentThreadIdCallback: () -> ThreadId,
+): BaseObjectTracker() {
```
Comment (eupp) [inline lincheck/src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/eventstructure/EventStructureObjectTracker.kt]:
```suggestion
) : BaseObjectTracker() {
```

File: lincheck/src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/eventstructure/EventStructure.kt [hunk 11]
```diff
@@ -34,6 +34,7 @@ internal class EventStructure(
     val memoryInitializer: MemoryInitializer,
     // TODO: refactor --- avoid using callbacks!
     private val reportInconsistencyCallback: ReportInconsistencyCallback,
+    private val currentThreadIdCallback: () -> ThreadId,
```
Comment (eupp) [inline lincheck/src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/eventstructure/EventStructure.kt]: Please add a `typealias` for the new callback, similarly as it is done for others.

File: lincheck/src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/eventstructure/EventStructureObjectTracker.kt [hunk 12]
```diff
@@ -94,7 +98,12 @@ internal class EventStructureObjectTracker(private val eventStructure: EventStru
                 allocation = initEvent!!
             )
         } else {
-            val iThread = (Thread.currentThread() as? TestThread)?.threadId ?: eventStructure.mainThreadId
+            // NOTE: currently, some for some parts of the code a thread id of -1 represents an invalid threadID value.
```
Comment (eupp) [inline lincheck/src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/eventstructure/EventStructureObjectTracker.kt]:
```suggestion
            // NOTE: currently, a thread id of -1 represents an invalid threadID value.
```

File: lincheck/src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt [hunk 13]
```diff
@@ -2708,6 +2715,9 @@ private fun TracePoint.isActorMethodCallTracePoint() =
 // represents an unknown code location
 internal const val UNKNOWN_CODE_LOCATION = -1
 
+// currently the exact place of System.arraycopy is not known
+internal const val SYSTEM_ARRAYCOPY_CODE_LOCATION = -2
```
Comment (eupp) [inline lincheck/src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt]: Just use `UNKNOWN_CODE_LOCATION`. No need to introduce special kind of code locations just for array copy.

File: lincheck/src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/MemoryLocation.kt [hunk 14]
```diff
@@ -183,7 +183,6 @@ class StaticFieldMemoryLocation(
         return (other is StaticFieldMemoryLocation)
                 && (className == other.className)
                 && (fieldName == other.fieldName)
-                && (kClass == other.kClass)
```
Comment (eupp) [inline lincheck/src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/MemoryLocation.kt:186]: Why `kClass` was removed from comparison?

File: lincheck/src/jvm/main/org/jetbrains/lincheck/datastructures/ModelCheckingOptions.kt [hunk 15]
```diff
@@ -37,6 +37,10 @@ class ModelCheckingOptions : ManagedOptions<ModelCheckingOptions, ModelCheckingC
         return this
     }
 
+    val isExperimentalModelCheckingEnabled: Boolean
```
Comment (eupp) [inline lincheck/src/jvm/main/org/jetbrains/lincheck/datastructures/ModelCheckingOptions.kt]:
```suggestion
    internal val isExperimentalModelCheckingEnabled: Boolean
```

File: lincheck/src/jvm/main/org/jetbrains/lincheck/Lincheck.kt [hunk 16]
```diff
@@ -46,8 +46,9 @@ object Lincheck {
     @JvmOverloads
     fun runConcurrentTest(
         invocations: Int = DEFAULT_INVOCATIONS,
+        useExperimentalModelChecking: Boolean = false,
         block: Runnable
-    ) = runConcurrentTestInternal(invocations, LincheckSettings.DEFAULT, block)
+    ) = runConcurrentTestInternal(invocations, useExperimentalModelChecking, LincheckSettings.DEFAULT, block)
```
Comment (eupp) [inline lincheck/src/jvm/main/org/jetbrains/lincheck/Lincheck.kt]: Please add flag `useExperimentalModelChecking` into `LincheckSettings` class instead of adding new argument.

File: lincheck/src/jvm/test/org/jetbrains/kotlinx/lincheck_test/strategy/eventstructure/PrimitivesTest.kt [hunk 17]
```diff
@@ -1467,6 +1470,514 @@ class PrimitivesTest {
...
+                }
+                thread {
+                    actor(TestClass::t2)
+                }
+            }
+            post {
+                actor(TestClass::post)
+            }
+        }
+        val outcomes =  setOf(5,7)
```
Comment (eupp) [inline lincheck/src/jvm/test/org/jetbrains/kotlinx/lincheck_test/strategy/eventstructure/PrimitivesTest.kt]:
```suggestion
        val outcomes = setOf(5,7)
```

File: lincheck/src/jvm/test/org/jetbrains/kotlinx/lincheck_test/strategy/eventstructure/PrimitivesTest.kt (hunk 5 above)
Comment (eupp) [reply lincheck/src/jvm/test/org/jetbrains/kotlinx/lincheck_test/strategy/eventstructure/PrimitivesTest.kt]: Some copy-paste is fine for test code.

File: lincheck/src/jvm/test/org/jetbrains/kotlinx/lincheck_test/strategy/eventstructure/PrimitivesTest.kt [hunk 18]
```diff
@@ -1467,6 +1470,514 @@ class PrimitivesTest {
...
+        litmusTest(AtomicFuLongVariable::class.java, testScenario, outcomes) { results ->
+            val r1 = getValue<Long>(results.parallelResults[0][0]!!)
+            val r2 = getValue<Long>(results.parallelResults[1][0]!!)
+            r1 to r2
+        }
+    }
+
+
+    @Test
+    fun testGC() {
```
Comment (eupp) [inline lincheck/src/jvm/test/org/jetbrains/kotlinx/lincheck_test/strategy/eventstructure/PrimitivesTest.kt]: Let's extract `testGC` from `PrimitivesTest` into separate file.

File: lincheck/src/jvm/test/org/jetbrains/kotlinx/lincheck_test/strategy/eventstructure/PrimitivesTest.kt [hunk 19]
```diff
@@ -1467,6 +1470,514 @@ class PrimitivesTest {
...
+
+            t1.join()
+            t2.join()
+            t3.join()
+            r0
+        }
+    }
+
+    @Test
+    fun testArrayCopy() {
```
Comment (eupp) [inline lincheck/src/jvm/test/org/jetbrains/kotlinx/lincheck_test/strategy/eventstructure/PrimitivesTest.kt]: Let's extract array copy tests from `PrimitivesTest` into a separate file.

**Review by eupp — CHANGES_REQUESTED** [review]

File: lincheck/src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/MemoryLocation.kt (hunk 14 above)
Comment (viliamDokov) [reply lincheck/src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/MemoryLocation.kt:186]: Two reasons: 

1. `hashCode()` does not use kClass, so this combined with one more issue for array memory locations, caused an annoying to find bug.
2. It is also slow
3. (Extra) Could not think of a good reason to keep it. In theory it the tuple of `(className, fieldName)` uniquely determines the kClass as, there cannot be two fields with the same name but a different type. However, I can miss some edge case I am not aware of...

File: lincheck/src/jvm/test-jdk11/org/jetbrains/kotlinx/lincheck_test/strategy/eventstructure/VarHandleTests.kt (hunk 4 above)
Comment (viliamDokov) [reply lincheck/src/jvm/test-jdk11/org/jetbrains/kotlinx/lincheck_test/strategy/eventstructure/VarHandleTests.kt:217]: It does

File: bootstrap/src/sun/nio/ch/lincheck/EventTracker.java (hunk 6 above)
Comment (viliamDokov) [reply bootstrap/src/sun/nio/ch/lincheck/EventTracker.java]: Ok, I can move `System.arraycopy`, but what about `Array.newInstance`? Maybe the `visitMethodInsn` of `ObjectCreationTransformer` is a good place for it.

Comment (viliamDokov) [reply bootstrap/src/sun/nio/ch/lincheck/EventTracker.java]: There is also a TODO in `processMethodEffectOnStaticSnapshot` for arraycopy and reflection. Should I fill in some sort of implementation there as well, or leave it for a later PR?

Comment (eupp) [reply bootstrap/src/sun/nio/ch/lincheck/EventTracker.java]: > Ok, I can move `System.arraycopy`, but what about `Array.newInstance`? Maybe the `visitMethodInsn` of `ObjectCreationTransformer` is a good place for it.

Yeah, putting it into `ObjectCreationTransformer` is a good idea. 

> There is also a TODO in `processMethodEffectOnStaticSnapshot` for arraycopy and reflection. Should I fill in some sort of implementation there as well, or leave it for a later PR?

If it's not hard to do, let's give some impl. Otherwise, I would leave as is for now.

Comment (viliamDokov) [reply bootstrap/src/sun/nio/ch/lincheck/EventTracker.java]: Actually looking at it right now, maybe it is better to lump in Array.newInstance with `processIntrinsicMethodEffects`? 
It seems like a cleaner option, compared to the gymnastics that we do right now in `ObjectCreationTransformer`? 
Not sure what you think, maybe it does not really matter in the end.

File: jvm-agent/src/main/org/jetbrains/lincheck/jvm/agent/LincheckClassVisitor.kt [hunk 20]
```diff
@@ -239,7 +239,7 @@ internal class LincheckClassVisitor(
         }
 
         // Must appear last in the code, to completely hide intrinsic candidate methods from all transformers
-        if (instrumentationMode == MODEL_CHECKING) {
+        if (instrumentationMode == MODEL_CHECKING || instrumentationMode == EXPERIMENTAL_MODEL_CHECKING) {
```
Comment (viliamDokov) [inline jvm-agent/src/main/org/jetbrains/lincheck/jvm/agent/LincheckClassVisitor.kt:242]: This was the change that fixed the CI fails for JDK8 and JDK11.

The other fails are some strange timeouts, which I could not reproduce after some light attempts at reproduction. Maybe we can run CI again to see what happens?

File: bootstrap/src/sun/nio/ch/lincheck/EventTracker.java (hunk 6 above)
Comment (eupp) [reply bootstrap/src/sun/nio/ch/lincheck/EventTracker.java]: Let's keep it in `ObjectCreationTransformer` for now.

File: integration-test/lincheck/src/main/org/jetbrains/lincheck_test/gpmc/CountDownLatchTest.kt [hunk 21]
```diff
@@ -11,28 +11,35 @@
...
+            counter.incrementAndGet()
+        }
 
         latch.countDown()
         threads.forEach { it.join() }
         check(counter.get() == nThreads)
     }
+
+    @Test
+    fun testLatchCountdown() = Lincheck.runConcurrentTest(10000,  block)
```
Comment (eupp) [inline integration-test/lincheck/src/main/org/jetbrains/lincheck_test/gpmc/CountDownLatchTest.kt]:
```suggestion
    fun testLatchCountdown() = Lincheck.runConcurrentTest(10000, block)
```

File: integration-test/lincheck/src/main/org/jetbrains/lincheck_test/gpmc/CountDownLatchTest.kt [hunk 22]
```diff
@@ -11,28 +11,35 @@
...
+
+    @Ignore("Times out")
+    @Test
+    fun testLatchCountdownEventStructure() = Lincheck.runConcurrentTest(10000,  block)
```
Comment (eupp) [inline integration-test/lincheck/src/main/org/jetbrains/lincheck_test/gpmc/CountDownLatchTest.kt]:
```suggestion
    fun testLatchCountdownEventStructure() = Lincheck.runConcurrentTest(10000, block)
```

Comment (eupp) [inline integration-test/lincheck/src/main/org/jetbrains/lincheck_test/gpmc/CountDownLatchTest.kt]: Looks like you forgot to add an option to run the test with event-structure strategy.

File: integration-test/lincheck/src/main/org/jetbrains/lincheck_test/gpmc/CyclicBarrierTest.kt [hunk 23]
```diff
@@ -31,4 +31,11 @@ class CyclicBarrierTest {
         t1.join()
         t2.join()
     }
+
+    @Test
+    fun testBarrier() = runConcurrentTest(10000,  block)
```
Comment (eupp) [inline integration-test/lincheck/src/main/org/jetbrains/lincheck_test/gpmc/CyclicBarrierTest.kt]:
```suggestion
    fun testBarrier() = runConcurrentTest(10000, block)
```

File: integration-test/lincheck/src/main/org/jetbrains/lincheck_test/gpmc/CyclicBarrierTest.kt [hunk 24]
```diff
@@ -31,4 +31,11 @@ class CyclicBarrierTest {
...
+
+    @Ignore("Times out")
+    @Test
+    fun testBarrierEventStructure() = runConcurrentTest(10000,  block)
```
Comment (eupp) [inline integration-test/lincheck/src/main/org/jetbrains/lincheck_test/gpmc/CyclicBarrierTest.kt]:
```suggestion
    fun testBarrierEventStructure() = runConcurrentTest(10000, block)
```

Comment (eupp) [inline integration-test/lincheck/src/main/org/jetbrains/lincheck_test/gpmc/CyclicBarrierTest.kt]: Looks like you forgot to add an option to run the test with event-structure strategy.

File: jvm-agent/src/main/org/jetbrains/lincheck/jvm/agent/transformers/ObjectCreationTransformer.kt [hunk 25]
```diff
@@ -350,4 +359,81 @@ internal class ObjectCreationTransformer(
         //   to avoid the runtime overhead when immutable-value tracking is disabled.
         bootstrapMethodOwner == "java/lang/invoke/StringConcatFactory"
 
+
+    private fun visitInvokeArrayNewInstance(opcode: Int, owner: String, name: String, descriptor: String, isInterface: Boolean) = adapter.run {
+        // TODO: should also call beforeNewObjectCreation?
```
Comment (eupp) [inline jvm-agent/src/main/org/jetbrains/lincheck/jvm/agent/transformers/ObjectCreationTransformer.kt]: Let's try to do this. If tests pass, let's keep `beforeNewObjectCreation`.

**Review by eupp — APPROVED** [review]
